api/controllers/export-data-2.js

```diff
@@ -0,0 +1,257 @@
+const _ = require('underscore');
+
+const PouchDB = require('pouchdb-core');
+PouchDB.plugin(require('pouchdb-adapter-http'));
+PouchDB.plugin(require('pouchdb-mapreduce'));
+
+const DB = new PouchDB(process.env.COUCH_URL);
+
+const { Readable } = require('stream'),
+      search = require('search')(Promise, DB);
+
+const BATCH = 100;
+
+const JOIN_COL = ',';
+const JOIN_ROW = '\n';
+
+const SUPPORTED_EXPORTS = ['reports'];
+
+
+//
+// TODO: these functions is copied from `export-data.js`, and modified to use
+// PouchDB / promises. We should remove both copies of these functions and create
+// a shared lineage library that we can use here, as well as in the rest of api,
+// sentinel and webapp.
+//
+// lineage.js in sentinel is probably the most comprehensive lineage library
+// right now, though it would need to be ES5-a-fied if we wanted to use it on
+// the front end.
+//
+// https://github.com/medic/medic-webapp/issues/XXXX
+//
+const findContact = (contactRows, id) => {
+  return id && contactRows.find(contactRow => contactRow.id === id);
+};
+const hydrateDataRecords = result => {
+  const rows = result.rows;
+
+  const contactIds = [];
+  rows.forEach(row => {
+    let parent = row.doc.contact;
+    while(parent) {
+      if (parent._id) {
+        contactIds.push(parent._id);
+      }
+      parent = parent.parent;
+    }
+  });
+
+  if (!contactIds.length) {
+    return Promise.resolve(result);
+  }
+
+  return DB.allDocs({
+    keys: contactIds,
+    include_docs: true
+  }).then(results => {
+    rows.forEach(row => {
+      const contactId = row.doc.contact && row.doc.contact._id;
+      const contact = findContact(results.rows, contactId);
+      if (contact) {
+        row.doc.contact = contact.doc;
+      }
+      let parent = row.doc.contact;
+      while(parent) {
+        const parentId = parent.parent && parent.parent._id;
+        const found = findContact(results.rows, parentId);
+        if (found) {
+          parent.parent = found.doc;
+        }
+        parent = parent.parent;
+      }
+    });
+
+    return result;
+  });
+};
+
+
+/**
+ * Flattens a given object into an object where the keys are dot-notation
+ * paths to the flattened values:
+ * {
+ *   "foo": {
+ *     "bar": "smang"
+ *   }
+ * }
+ *
+ * becomes:
+ *
+ * {
+ *   "foo.bar": "smang"
+ * }
+ */
+// TODO: is there really not a utility that does this?
+const flatten = (fields, prepend=[]) =>
+  Object.keys(fields).reduce((acc, k) => {
+    const path = [...prepend, k];
+    if (typeof fields[k] === 'object' && fields[k]) {
+      return _.extend(acc, flatten(fields[k], path));
+    } else {
+      acc[path.join('.')] = fields[k] || '';
+      return acc;
+    }
+  }, {});
```

Comment:
Could you take a stab at refactoring this in a way you thing looks good? I'm not really sure how to change this, apart from:

```
const flatten = (fields, prepend=[]) => {
  return Object.keys(fields).reduce((acc, k) => {
    const path = [...prepend, k];
    if (typeof fields[k] === 'object' && fields[k]) {
      return _.extend(acc, flatten(fields[k], path));
    } else {
      acc[path.join('.')] = fields[k];
      return acc;
    }
  }, {});
};
```

Which, at least from my perspective, doesn't add any clarity.

Reply:
That's exactly what I meant. It adds clarity for me, because without that it looks like the function doesn't return anything. I had to read it a couple of times to spot the missing braces.

If it were me writing it I'd also move the return statement to the bottom of the reduce function to make it clearer that `acc` is always returned, eg:

```js
const flatten = (fields, prepend=[]) => {
  return Object.keys(fields).reduce((acc, k) => {
    const path = [...prepend, k];
    if (typeof fields[k] === 'object' && fields[k]) {
      _.extend(acc, flatten(fields[k], path));
    } else {
      acc[path.join('.')] = fields[k];
    }
    return acc;
  }, {});
};
```